Require authentication for SCIM requests

NobodysNightmare opened the pull request:
Handling authentication through our regular warden strategies. Doing this required to make warden available to the Rails stack as well. This was not super straight-forward, because of load order issues:

* Requiring a Rails middleware must be done before initialization finished
* Our warden config was so far done _after_ initialization
* static_routes were defined in lib, which is automatically reloaded, but auto-reloading code is now allowed during initialization
    * lib_static which is autoloaded_once is fine during init

# Ticket
https://community.openproject.org/work_packages/62592/activity

# Notes

I'd like to avoid overwriting internal scimitar methods entirely. I opened an upstream issue in that regard, maybe this opens some ideas for API changes: https://github.com/pond/scimitar/issues/158

# Dependencies

* [x] https://github.com/opf/openproject/pull/19017 needs to be merged first

NobodysNightmare committed:
Require authentication for SCIM requests

Handling authentication through our regular warden strategies.

Permissions-wise the only thing we can check for is whether the authenticated
user is an admin. Though this will require us to grant admin privileges to
all SCIM clients, which might be more than we want to do.

NobodysNightmare committed:
Fix JWT OIDC strategy

ba1ash committed:
Merge branch 'feature/62107-add-scim-server-api' into scim-authentication

ba1ash commented on app/models/user.rb:
```diff
@@ -771,3 +771,7 @@ def self.default_admin_account_changed?
+
+# Fix for development (and non eagerloaded environments), where User.find(...) would not return service accounts before
+# they've been autoloaded.
+require "service_account"
```
@NobodysNightmare 
It seems this statement breaks tests. I suggest to resolve this issue a bit differently in https://github.com/opf/openproject/pull/19266. Kindly have a look.

ba1ash committed:
Merge branch 'feature/62107-add-scim-server-api' into scim-authentication

ba1ash committed:
Merge branch 'feature/62107-add-scim-server-api' into scim-authentication

ba1ash committed:
[#62107] Update SCIM Server API.

- Use authenticated ServiceAccount in requests
- Scope User and Group requests by ScimClient related auth_provider_id
- Include ServiceAccount search to doorkeeper_oauth strategy
- Fix SCIM Server API specs.

ba1ash committed:
[#62107] Update SCIM Server API.

- Use ServiceAccount associated with ScimClient for making user changes
- Remove scoping by scim_client.auth_provider_id
  So, SCIM Client has access to any not_builtin User.
- Associate user with AuthProvider configured in ScimModel
  instead of choosing the first one.

ba1ash committed:
[#62107] Fix specs. Add openjdk to backend image.

- Add openjdk to backend image.
  It is required to sping up test LDAP server in specs.
  https://github.com/NUARIG/ladle

ba1ash committed:
Merge branch 'feature/62107-add-scim-server-api' into scim-authentication
